simulator/core/state.h

```diff
@@ -19,19 +19,74 @@
 #include "pretty_print.h"
 #endif
 
+/** Return whether the debugging state is active.
+ *
+ * Debugging state is affected by state_enter_debugging() and
+ * state_exit_debugging().
+ *
+ * @return True if debugging is active, false otherwise.
+ */
 bool state_is_debugging(void);
+
+/** Start/reset the write state tracking.
+ *
+ * Internally, it merely resets the index to the write state tracker to 0.
+ *
+ * If replay is enabled, this manages the linked list of states. It adds the
+ * current state to the linked list, invalidating any states that are newer
+ * than the current one (if there are any, this happens if we're re-executing
+ * after rewinding). The current cycle count is recorded in the state.
+ */
 void state_start_tick(void);
+
+/** Update the simulator state based on the current write state.
+ *
+ * Specficially, it goes through all write states that have not been committed
+ * simce the last start_start_tick call and writes the value specified in the
+ * write state to the specified location (there are two variants of this-- if
+ * state_change.loc is set, the state_change.val variable is written to
+ * *state_change.loc, else the pointer at state_change.pval is written to the
+ * location pointed to by state_change.ploc).
+ */
 void state_tock(void);
+
 #ifndef NO_PIPELINE
+/** Signal the emulator to start flushing the pipeline.
+ *
+ * This sets a flag that the thread managing the pipeline watches for to start
+ * flushing the pipeline.
+ *
+ * @param[in] new_pc The new PC value to set the simulator to once flushing is
+ *  completed.
+ */
 void state_pipeline_flush(uint32_t new_pc);
 #endif
 
+/** FIXME This isn't implemented anywhere?
+ *
+ */
 void state_async_block_start(void);
+
+/** FIXME This isn't implemented anywhere?
+ *
+ */
 void state_async_block_end(void);
 
+/** Set the emulator's debug flag to true.
+ */
 void state_enter_debugging(void);
+
+/** Set the emulator's debug flag to false.
+ */
 void state_exit_debugging(void);
```

Comment:
It may be worth noting that for state tracking, debugging is slightly more significant. In particular, it allows writes to proceed to directly affect the underlying hardware representation, rather than going in the tick/tock event queue.